kvdb-web/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+		match columns.columns.get(column) {
+			None => Err(io::Error::new(io::ErrorKind::Other, format!("No such column family: {:?}", col))),
+			Some(map) => Ok(map.get(key).cloned().map(|v| v.inner).map(DBValue::from_vec)),
+		}
+	}
+
+	fn get_by_prefix(&self, col: Option<u32>, prefix: &[u8]) -> Option<Box<[u8]>> {
+		let col = Self::to_db_column(col);
+		let columns = self.columns.read().expect_throw(RWLOCK_NO_POISONING_PROOF);
+		match columns.columns.get(col) {
+			None => None,
+			Some(map) =>
+				map.iter().find(|&(ref k ,_)| k.starts_with(prefix))
+					.map(|(_, v)| v.inner.clone().into_boxed_slice())
+		}
+	}
+
+	fn write_buffered(&self, transaction: DBTransaction) {
+		let ops = transaction.ops;
+		let mut columns = self.columns.write().expect_throw(RWLOCK_NO_POISONING_PROOF);
+		for op in ops {
+			match op {
+				DBOp::Insert { col, key, value } => {
+					let col = Self::to_db_column(col);
+					if let Some(col) = columns.columns.get_mut(col) {
+						col.insert(
+							key.into_vec().into(),
+							value.into_vec().into(),
+						);
+					}
+				},
+				DBOp::Delete { col, key } => {
+					let col = Self::to_db_column(col);
+					if let Some(col) = columns.columns.get_mut(col) {
+						col.remove(&*key);
+					}
+				},
+			}
+		}
+	}
+
+	fn flush(&self) -> io::Result<()> {
+		Ok(())
+	}
+
+	// NOTE: clones the whole db
+	fn iter<'a>(&'a self, col: Option<u32>) -> Box<Iterator<Item=(Box<[u8]>, Box<[u8]>)> + 'a> {
+		let col = Self::to_db_column(col);
+		let columns = self.columns.read().expect_throw(RWLOCK_NO_POISONING_PROOF);
+		match columns.columns.get(col) {
+			Some(map) => Box::new(
+				map.clone()
+					.into_iter()
+					.map(|(k, v)| (k.inner.into_boxed_slice(), v.inner.into_boxed_slice()))
+			),
+			None => Box::new(None.into_iter()),
+		}
+	}
+
+	// NOTE: clones the whole db
+	fn iter_from_prefix<'a>(&'a self, col: Option<u32>, prefix: &'a [u8])
+		-> Box<Iterator<Item=(Box<[u8]>, Box<[u8]>)> + 'a>
+	{
+		let col = Self::to_db_column(col);
+		let columns = self.columns.read().expect_throw(RWLOCK_NO_POISONING_PROOF);
+		match columns.columns.get(col) {
+			Some(map) => Box::new(
+				map.clone()
+					.into_iter()
+					.skip_while(move |&(ref k, _)| !k.starts_with(prefix))
+					.map(|(k, v)| (k.inner.into_boxed_slice(), v.inner.into_boxed_slice()))
+			),
+			None => Box::new(None.into_iter()),
+		}
+	}
+
+	// NOTE: not supported
+	fn restore(&self, _new_db: &str) -> std::io::Result<()> {
+		Err(io::Error::new(io::ErrorKind::Other, "Attempted to restore an in-browser database"))
+	}
+}
+
+
+fn persist_db(db: Arc<Database>) {
+	const PERSIST_DB_TIMER_MS: i32 = 2_000;
+
+	let window = web_sys::window().expect_throw("are we in a browser?");
+	let local_storage = window.local_storage()
+		.expect_throw("localStorage should be supported in your browser")
+		.unwrap_throw();
+
+	let a = Closure::wrap(Box::new(move || do_persist_db(&db.clone(), &local_storage)) as Box<dyn Fn()>);
```

Comment:
We should use a `Weak<Database>` here, so that if we close the database, the flushing stops as well.

Ideally we would also stop the timer. This can be done by using `set_timeout` instead of `set_interval`, and having the callback call `set_timeout` again, unless the database is dead.
I'm not sure of the difficulty of doing that, and I'd fine to merge with `set_interval`.